src/Files.App/Filesystem/Security/AccessControlList.cs

```diff
@@ -0,0 +1,255 @@
+		}
+
+		public FileSystemRights GetEffectiveRights()
+		{
+			using var user = WindowsIdentity.GetCurrent();
+			var userSids = new List<string> { user.User.Value };
+			userSids.AddRange(user.Groups.Select(x => x.Value));
+
+			FileSystemRights inheritedDenyRights = 0, denyRights = 0;
+			FileSystemRights inheritedAllowRights = 0, allowRights = 0;
+
+			foreach (var Rule in AccessControlEntryPrimitiveMappingList.Where(x => userSids.Contains(x.PrincipalSid)))
+			{
+				if (Rule.AccessControlType == System.Security.AccessControl.AccessControlType.Deny)
+				{
+					if (Rule.IsInherited)
+					{
+						inheritedDenyRights |= Rule.FileSystemRights;
+					}
+					else
+					{
+						denyRights |= Rule.FileSystemRights;
+					}
+				}
+				else if (Rule.AccessControlType == System.Security.AccessControl.AccessControlType.Allow)
+				{
+					if (Rule.IsInherited)
+					{
+						inheritedAllowRights |= Rule.FileSystemRights;
+					}
+					else
+					{
+						allowRights |= Rule.FileSystemRights;
+					}
+				}
+			}
+
+			return (inheritedAllowRights & ~inheritedDenyRights) | (allowRights & ~denyRights);
+		}
+
+		public static AccessControlList FromPath(string path, bool isFolder)
+		{
+			var filePermissions = new AccessControlList()
+			{
+				Path = path,
+				IsFolder = isFolder
+			};
+
+			var acsResult = GetAccessControl(path, isFolder, out var acs);
+			if (acsResult)
+			{
+				var rules = new List<AccessControlEntryPrimitiveMapping>();
+
+				var accessRules = acs.GetAccessRules(true, true, typeof(SecurityIdentifier));
+				foreach (var accessRule in accessRules)
+				{
+					rules.Add(AccessControlEntryPrimitiveMapping.FromFileSystemAccessRule((FileSystemAccessRule)accessRule));
+				}
+
+				filePermissions.AccessControlEntryPrimitiveMappingList.AddRange(rules);
+				filePermissions.OwnerSID = acs.GetOwner(typeof(SecurityIdentifier)).Value;
+				filePermissions.IsAccessControlListProtected = acs.AreAccessRulesProtected;
+			}
+
+			filePermissions.CanReadAccessControl = acsResult;
+
+			return filePermissions;
+		}
+
+		private static bool GetAccessControl(string path, bool isFolder, out FileSystemSecurity fss)
+		{
+			try
+			{
+				if (isFolder && Directory.Exists(path))
+				{
+					fss = FileSystemAclExtensions.GetAccessControl(new DirectoryInfo(path));
+
+					return true;
+				}
+				else if (File.Exists(path))
+				{
+					fss = FileSystemAclExtensions.GetAccessControl(new FileInfo(path));
+
+					return true;
+				}
+				else
+				{
+					// File or folder does not exists
+					fss = null;
+
+					return false;
+				}
+			}
+			catch (UnauthorizedAccessException)
+			{
+				// User does not have rights to read access rules
+				fss = null;
+
+				return false;
+			}
+			catch (Exception)
+			{
+				fss = null;
+
+				return false;
+			}
```

Comment:
You can move `fss = null; return false;` to the end of the method, this way we would avoid code duplication